### PR TITLE
Implement method to create File based on a buffer

### LINKED ISSUE #17: Missing method to create a fake awmp::File
Right now I have a problem where I need to create a fake awmp::File for testing purposes. I propose to create a new method in `File` to solve this problem. As far as I'm concerned, the implementation itself should be trivial and I would love to make a PR for it.

If there's any other solution or feedback regarding this problem, I would love to hear it.

### PR DESCRIPTION
**Pull request by tengkuizdihar**
resolve #17 

### Note
* The reason why `create_from_buffer` and `create_with_size` take an optional `temp_dir` argument was because of the notes that I've made 5 months ago about how temporary file are stored. In which the temporary file need to be in the same file system where it will be persisted in. For example, using the default temp file directory, it will save the temp file in `/run/user/1000` and in some UNIX system that I've used, this will cause a disturbance when I want to persist it in `/home/user/myapplication` for example.

### CODE REVIEW

**Comment by kardeiz** [comment]
Hi, thanks for the PR! I would like to request a few changes:

Let's re-export tempfile's `NamedTempFile` as discussed in the issue. I.e. in lib.rs:

```rust
/// Re-export of `tempfile::NamedTempFile`
pub use tempfile::NamedTempFile;
```

and then change the constructor to take `(original_file_name: Option<String>, file: NamedTempFile)`. Please use https://github.com/kardeiz/awmp/blob/main/src/v2_3.rs#L152-L168 for reference. I'll make a few other comments in-line.

**EDIT**

Actually, let's have the constructor be:

```rust
pub fn new(file: NamedTempFile, original_file_name: Option<String>, mime_type: Option<&mime::Mime>) -> Self {
    let sanitized_file_name = match original_file_name {
        Some(ref s) => sanitize_filename::sanitize(s),
        None => {
            let uuid = uuid::Uuid::new_v4().to_simple();
            match mime_type.and_then(|mt| mime_guess::get_mime_extensions(mt)).and_then(|x| x.first()) {
                Some(ext) => format!("{}.{}", uuid, ext),
                None => uuid.to_string(),
            }
        }
    };

    File {
        inner: file,
        sanitized_file_name,
        original_file_name,
    }
}

pub fn new_with_file_name(file: NamedTempFile, original_file_name: String) -> Self {
    Self::new(file, Some(original_file_name), None)
}
```

Then https://github.com/kardeiz/awmp/blob/main/src/v2_3.rs#L152-L168 can be changed to look like:

```rust
(None, Buffer::File(file)) => {
    Ok((
        name,
        Part::File(Ok(File::new(file, file_name_opt, Some(&mime_type)))),
    ))
}
```

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,6 @@
 /target
 **/*.rs.bk
 Cargo.lock
+
+# File made from tests
+create_file_then_persist.txt
```
Comment (kardeiz) [inline .gitignore]: Please remove this, and modify the test to create the tempfile in a temp directory (outside the crate).

File: src/lib.rs [hunk 2]
```diff
@@ -275,20 +275,44 @@ impl File {
...
     /// Also neither the file contents nor the containing directory are
     /// synchronized, so the update may not yet have reached the disk when
     /// `persist_at` returns.
     pub fn persist_at<P: AsRef<Path>>(self, path: P) -> Result<std::fs::File, Error> {
         self.inner.persist(path).map_err(Error::TempFilePersistError)
     }
+
+    /// Create a File from the given buffer.
+    /// Could be used in testing scenarios.
+    pub fn create_from_buffer<P: AsRef<Path>>(buf: &[u8], original_file_name: String, temp_dir: Option<P>) -> Result<Self, Error> {
```
Comment (kardeiz) [inline src/lib.rs]: Let's call this `new` (or `create`), and please see my other comments for the new signature.

File: src/lib.rs [hunk 3]
```diff
@@ -275,20 +275,44 @@ impl File {
...
+        Ok(Self {
+            inner: file,
+            original_file_name: Some(original_file_name.clone()),
+            sanitized_file_name: sanitize_filename::sanitize(original_file_name),
+        })
+    }
+
+    /// Create a File with a given size.
+    /// Could be used in testing scenarios.
+    pub fn create_with_size<P: AsRef<Path>>(size_bytes: usize, original_file_name: String, temp_dir: Option<P>) -> Result<Self, Error> {
```
Comment (kardeiz) [inline src/lib.rs]: Please remove this function entirely; I don't think this library needs to provide this functionality. Users can easily do this outside the crate.

**Review by kardeiz — CHANGES_REQUESTED** [review]

**Comment by tengkuizdihar** [comment]
Thank you for the review, I will fix it soon

File: .gitignore (hunk 1 above)
Comment (tengkuizdihar) [reply .gitignore]: make sense. ill do that. i did this because i dont know which approach to use.

**Comment by tengkuizdihar** [comment]
It should be done and its modified according to your review
thanks man, i learn something about rust and mime types!

**Comment by tengkuizdihar** [comment]
I've updated the PR based on your review, I hope this would be enough to be merged.

PS, what do you think about the clippy warnings? Should the project adhere to that?

**Review by kardeiz — APPROVED** [review]
Looks good to me! Thanks!

**Comment by kardeiz** [comment]
I've just published this as `0.7.0`. Thanks again! Thanks for reminding me about clippy; I went ahead and fixed most of the warnings.

**Pull request merged**